lib/filter/filter-throttle.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2021 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filter-throttle.h"
+#include "timeutils/misc.h"
+#include "scratch-buffers.h"
+#include "str-utils.h"
+
+typedef struct _FilterThrottle
+{
+  FilterExprNode super;
+  NVHandle key_handle;
+  gint rate;
+  GMutex *map_lock;
+  GHashTable *rate_limits;
+} FilterThrottle;
+
+typedef struct _ThrottleRateLimit
+{
+  gint tokens;
+  GTimeVal last_check;
+  GMutex *ratelimit_lock;
+} ThrottleRateLimit;
+
+static ThrottleRateLimit *
+throttle_ratelimit_new()
+{
+  ThrottleRateLimit *self = g_new0(ThrottleRateLimit, 1);
+  self->ratelimit_lock = g_mutex_new();
+  return self;
+}
+
+static void
+throttle_ratelimit_free(ThrottleRateLimit *self)
+{
+  g_mutex_free(self->ratelimit_lock);
+  g_free(self);
+}
+
+static void
+throttle_ratelimit_add_new_tokens(ThrottleRateLimit *self, gint rate)
+{
+  glong diff;
+  gint num_new_tokens;
+  GTimeVal now;
+  g_get_current_time(&now);
+
+  if (self->last_check.tv_sec == 0)
+    {
+      self->last_check = now;
+      self->tokens = rate;
+      diff = 0;
+    }
```

Comment:
I think my thinking at the time was that if we just initialised it, we could short circuit the new tokens calculation (but didn't end up doing that 🙈). Moving it to the constructor makes sense though, as does storing the rate in the ThrottleRateLimit. I don't think that the shortcircuiting logic there would have saved us much anyway - it would only be on the very first initialisation, at the cost of a branch in every eval

Reply:
I think the only qualm I have about this is that we duplicate the

```
  GTimeVal now;
  g_get_current_time(&now);
```

logic, but I guess that's not exactly a large snippet to copy

Reply:
Thanks! Yes, I think we can live that code duplication. :)